adding dns support for dnspod provider

piratimir opened the pull request:
## DNS support for DNSPod provider
### Description

I coded the dns driver for DNSPod on libcloud/dns/drivers/dnspod.py module. In order for this driver to work the libcloud/common/dnspod.py module is needed. The unitttests for this driver are coded in the libcloud/test/dns/test_dnspod.py module.
### Status
- done, ready for review
### Checklist (tick everything that applies)
- [ ] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [ ] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

Kami commented:
Great, thanks.

Is the PR already for a review, or are you still working on it?

piratimir commented:
@Kami the pr is ready for review :)

piratimir committed:
adding dns support for dnspod provider

piratimir committed:
updates the secrets file

piratimir committed:
updates the driver

piratimir committed:
updates code to pep8 standards, uses urlencode to support different python
versions

piratimir committed:
removes unused import statements such as import urllib

piratimir committed:
fixes the lint for test_dnspod module

Kami commented on libcloud/test/dns/test_dnspod.py:
```diff
@@ -0,0 +1,266 @@
+                                                           url, body, headers):
+        body = self.fixtures.load('delete_record_record_does_not_exist.json')
+
+        return httplib.OK, body, {}, httplib.responses[httplib.OK]
```
Just curious - does API indeed return 200 status code for all the different responses?

Kami commented on libcloud/common/dnspod.py:
```diff
@@ -0,0 +1,65 @@
+from libcloud.common.base import ConnectionKey, JsonResponse
```
Please add a missing Apache license header.

Kami commented on libcloud/common/dnspod.py:
```diff
@@ -0,0 +1,65 @@
+    def add_default_headers(self, headers):
+        headers['Content-Type'] = 'application/x-www-form-urlencoded'
+        headers['Accept'] = 'text/json'
+        headers['User-Agent'] = \
```
Please remove this, libcloud includes it's own custom User-Agent header (it includes version of Libcloud and driver user).

Kami commented on libcloud/dns/drivers/dnspod.py:
```diff
@@ -0,0 +1,345 @@
+
+    def list_zones(self):
+        action = '/Domain.List'
+        data = {'user_token': self.key, 'format': 'json'}
```
`user_token` and `format` attributes seem to be included in every request so you could maybe refactor this a bit and include those attributes somewhere else (e.g. inside the connection `request` method or similar) so the same piece of code doesn't need to be repeated all over.

piratimir replied:
@Kami yes i thought about that, how to include that in every request?

piratimir replied:
@Kami hey man any update on this? I have fixed the other problems, this is the only one left.

Kami replied:
As mentioned in my comment above, you could refactor that functionality in a custom `request` method on the connection class.

You can find a similar example here - https://github.com/apache/libcloud/blob/trunk/libcloud/dns/drivers/cloudflare.py#L106

Kami commented on libcloud/dns/drivers/dnspod.py:
```diff
@@ -0,0 +1,345 @@
+            e = sys.exc_info()[1]
+            if e.message == 'No domains':
+                return []
+        zones = self._to_zones(items=response.parse_body()['domains'])
```
Libcloud already parses the body for you and is available via `response.object` attribute.

No need to call `parse_body` yourself.

Kami commented on libcloud/dns/drivers/dnspod.py:
```diff
@@ -0,0 +1,345 @@
+                                               data=data)
+        except DNSPodException:
+            e = sys.exc_info()[1]
+            if e.message in ['Domain is exists',
```
You could save those messages in a constant, e.g. `ZONE_ALREADY_EXISTS_ERROR_MSGS = [ ... ]`.

Kami commented:
Thanks :+1: 

I've added some in-line comments.

Kami commented:
@jetbird Did you push your latest changes?

Since I don't see any new commits (license header is still missing, etc.).

piratimir commented:
@Kami  done :)

piratimir committed:
places error messages in lists, fixes the duplicate problem data param
creates, removes uneeded default headers, calling response.object instead
of response.parse_body

Kami commented:
@jetbird Thanks.

I've made some changes and fixes (6594a46c53ef6d9953b742826c4fc5d86bc0ed58, defff8acd1028f78b58df32e8dcd179f728aaf0c) and merged patch into trunk.

As you can see in the second commit (defff8acd1028f78b58df32e8dcd179f728aaf0c), `_make_request` method didn't actually do anything with the `method` argument (a bug which existing tests don't catch). That was masked by (ab)using `**kwargs` - that's a good example of why `**kwargs` should be avoided in scenarios like that and only used when there is a valid need for it.

Kami commented:
@jetbird Can you please test latest changes against the live provider API?

piratimir commented:
Thanks for explaining the fixes :) Of course, I will test it now.

piratimir commented:
@Kami Tested. It works fine for me :)

Kami commented:
@jetbird Great :+1: 

It would also be good if you can add some basic docs for the driver and contact the provider (tall them about Libcloud, driver, etc. and ask them if it's possible to mention it somewhere on the page / docs, etc.).
